cmd/standalone/main.go

```diff
@@ -119,8 +124,13 @@ func main() {
 			qOpts := new(queryApp.QueryOptions).InitFromViper(v)
 
 			startAgent(aOpts, cOpts, logger, metricsFactory)
-			startCollector(cOpts, spanWriter, logger, metricsFactory, samplingHandler)
-			startQuery(qOpts, spanReader, dependencyReader, logger, metricsFactory, mBldr)
+			collectorServer := startCollector(cOpts, spanWriter, logger, metricsFactory, samplingHandler, hc)
+			queryServer := startQuery(qOpts, spanReader, dependencyReader, logger, metricsFactory, mBldr, hc)
+			logger.Info("Starting jaeger-collector HTTP server", zap.Int("http-port", cOpts.CollectorHTTPPort))
+			logger.Info("Starting jaeger-query HTTP server", zap.Int("port", qOpts.Port))
+			go collectorServer()
+			go queryServer()
+			hc.Ready()
```

Comment:
Move goroutines to main from each start function, and start them in main then set Ready.
IMO, this will prevent the race issue. 

For 100% guarantees, can we move hc.Ready before goroutines like this? 
```go
hc.Ready()
go collectorServer()
go queryServer()
```

Reply:
I don't think returning functions like this provides any additional guarantees. I think it's fine to leave it as it was, i.e. call `startCollector()`, it might fail mid-way and crash the program, or it can spawn goroutine for ListenAndServe, which can fail as well and crush the program later. So there will be a possible time interval where HC==Ready, and then a crash, but I don't think it's critical for all-in-one.

Sorry I led you on the wrong path with the last comment. The fact that the goroutines call log.Fatal on error sort of take care of things.

Reply:
I wanted hc.Ready() to be called before collector/query goroutine call hc.Set(healthcheck.Unavailable) in ListenAndServe error.
I think this would prevent overriding Unavailable by hc.Ready() as you mentioned above coment.

Anyway, if the issue in all-in-one is not critical and fatal log on error is enough, I'll revert.

Reply:
Yes, I don't think it's critical sice even if the override happens the process will exit anyway.